File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        height(root);
        return res; 
    }
    int height(TreeNode* node){
        if(!node) return 0; 
        int lst = height(node->left);
        int rst = height(node->right);
        res = max(res,lst+rst);
        return 1+max(lst,rst);
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def dfs(self,node,res) -> int: 
        if node is None: 
            return 0 
        lst = self.dfs(node.left,res)
        rst = self.dfs(node.right,res)
        res[0] = max(res[0],lst+rst)
        return 1+max(lst,rst)
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int: 
        # get heigh of lst and rst 
        # at each point max(res,lst+rst) 
        # return max(lst,rst)
        res = [0]
        self.dfs(root,res)
        return res[0]